src/persistence.rs

```diff
@@ -95,23 +107,26 @@ pub mod sparse_matrix {
         }
 
         fn get_hash(&self, id: u32) -> i64 {
-            match self.id_2_hash.get(&id) {
+            match self.id_2_hash.get(id as usize) {
                 Some(value) => *value as i64,
                 None => -1i64,
             }
         }
 
         fn get_entity_counter(&self) -> u32 {
-            self.entity_count
+            self.id_2_hash.len() as u32
         }
 
-        fn update_entity_counter(&mut self, val: u32) {
-            self.entity_count = val;
-        }
-
-        fn add_hash_id(&mut self, hash: u64, id: u32) {
-            self.hash_2_id.insert(hash, id);
-            self.id_2_hash.insert(id, hash);
+        fn get_or_add_id_by_hash(&mut self, hash: u64) -> u32 {
```

Comment:
I wanted to split storage (persistence) and logic of sparse matrix construction. I would prefer to move this logic to sparse matrix. If we want to use some other storage such as LMDB then we just provide storage without any logic.

Reply:
I would consider get_or_add operation as an atomic operation that may be provided by any persistent storage. Some databases (not sure about lmdb) allow to implement this in a single operation (thus perform this operation in one and not two blocking calls).